### ISSUE TITLE
Kubernetes v1.22+ container runtime on Fedora CoreOS

### ISSUE THREAD

**Issue by dghubble** [coreos/fedora-coreos-tracker#767]
Kubernetes intends to drop drop support for docker-shim as a container runtime in v1.22. Currently, Fedora CoreOS `33.20210217.3.0` ships docker `19.03.13`. docker-shim remains the most stable, tested, available out-of-the-box runtime, but this will end soon. I'd like some kind of clarity on Fedora CoreOS's intentions, such as shipping a compatible `containerd` or `cri-o`. With Kubernetes cutting v1.22 alpha releases (likely pretty soon) in the time frame of Fedora 34, as Kubernetes distros we'll want to start evaluating and conformance testing the selection.

Overall, I'd like to know there is some plan. Ideally a documented one. Flatcar Linux has published their intentions to ship containerd in time ([docs](https://kinvolk.io/docs/flatcar-container-linux/latest/container-runtimes/switching-from-docker-to-containerd-for-kubernetes/)) for the deadline.

### cri-o

It looks like `cri-o` can only be installed by downloading an RPM directly and rpm-ostree installing it (unverified). https://github.com/coreos/fedora-coreos-tracker/issues/292#issuecomment-796998069. I have some maturity concerns about that. Is this really the recommended path? For a container optimized distro to not have a better path to getting a container runtime?

Releases are also pinned to Kubernetes versions and seem to lag by quite some time, so I have some velocity concerns (the runtime needs to be fairly stable, we roll forward when Kubernetes does, think hours).

### containerd

I haven't seen Fedora CoreOS plans to make this the runtime of choice.

**Comment by dghubble** [comment]
:eyes: cri-o doesn't list Fedora CoreOS as a supported OS and the only mention of how to hack it in is a year ago from @dustymabe 

https://github.com/cri-o/cri-o/blob/master/install.md
https://discussion.fedoraproject.org/t/installing-using-cri-o-on-fedora-coreos/15961/5

**Comment by lucab** [comment]
> [cri-o] Releases are also pinned to Kubernetes versions and seem to lag by quite some time

This is the main reason why cri-o is not shipped as part of the base FCOS. The daemon isn't a generic container runtime that can be freely upgraded by the OS, but it is instead interlocked with the higher level application/service (k8s control plane).

This strict interlock is an explicit design, highlighted at https://github.com/cri-o/cri-o. To that extent, for k8s distributions the whole "kubelet plus cri-o" is effectively a single "node runtime" component.

**Comment by rhatdan** [comment]
@mrunalp @haircommander PTAL

**Comment by rhatdan** [comment]
@mrunalp @haircommander Why isn't Fedora a supported platform for CRI-O? Is this just an oversite?

**Comment by haircommander** [comment]
fedora is listed https://github.com/cri-o/cri-o/blob/master/install.md#fedora-31-or-later it's just not in the table (which is both out of date and unneeded)

**Comment by haircommander** [comment]
> > [cri-o] Releases are also pinned to Kubernetes versions and seem to lag by quite some time
> 
> This is the main reason why cri-o is not shipped as part of the base FCOS. The daemon isn't a generic container runtime that can be freely upgraded by the OS, but it is instead interlocked with the higher level application/service (k8s control plane).
> 
> This strict interlock is an explicit design, highlighted at https://github.com/cri-o/cri-o. To that extent, for k8s distributions the whole "kubelet plus cri-o" is effectively a single "node runtime" component.

I agree with @lucab here, kubelet and cri-o should be installed together with matching versions. If there's a way we (the cri-o team) can minimize the friction with this in FCOS, I'd love to hear about it.

**Comment by LorbusChris** [comment]
As I see it, the issue here is more that it's a bit difficult to install the cri-o RPM because it's distributed as a module and rpm-ostree can't handle it.
At least for OKD, making the version of cri-o that is used in the current OKD version available in a standard, non-modular yum repo would help since we could then include it easily (either in the base compose, or as an extension in a local yum repo)
Longer term, rpm-ostree should probably be taught to work with modular repos.

**Comment by haircommander** [comment]
how is the kubelet packaged for OKD now? Could the appropriate cri-o version be tossed in that package?

**Comment by vrutkovs** [comment]
In OKD we grab `kubelet` RPM from [origin artifacts](https://github.com/openshift/okd-machine-os/blob/master/Dockerfile#L2), copy it to [`/tmp/rpms`](https://github.com/openshift/okd-machine-os/blob/master/Dockerfile#L6), [extract to /tmp/working](https://github.com/openshift/okd-machine-os/blob/master/entrypoint.sh#L140-L143) and [create a new rpm-ostree commit with this dir overlayed](https://github.com/openshift/okd-machine-os/blob/master/entrypoint.sh#L157).

Similar steps to [fetch CRI-O RPM](https://github.com/openshift/okd-machine-os/blob/master/entrypoint.sh#L134-L135)

**Comment by dghubble** [comment]
For comparison, docker+shim (and soon containerd on Flatcar Linux) provide a suitable container runtime that meets Kubernetes  CRI minimum needs, out of the box.  Kublet is bundled as a container [image](https://github.com/poseidon/kubelet). Runs. Conformant. We roll forward within days of Kubernetes releases and do need to not wait on changes from the base OS or any package ecosystem. 

If we move toward `cri-o`, in addition to the friction of adopting Fedora CoreOS (it feels like we'd be the first user outside okd), there is this velocity concern. I appreciate you folks have explicit RPM packages for components. In pratice, we'd have to release new Kubernetes with Flatcar Linux only, and somehow have Fedora CoreOS come weeks later or something. We'd be waiting on this package ecosystem. How strict is the `cri-o` to `k8s` versioning really? Is it just about valiation?

**Comment by haircommander** [comment]
> How strict is the cri-o to k8s versioning really? Is it just about valiation?

in practice, not terribly strict. It's the safest bet, though. the cri, while generally stable, does change. We don't backport new cri changes to older versions of cri-o. We also don't attempt to test any kubelet/cri-o skew. Basically: we make no support claims for anything other than matching versions. 

Generally, folks don't have much trouble with having mismatched versions (I've never heard anyone complain about it). But it's theoretically possible

>  Kublet is bundled as a container image.

out of curiosity, do you run the kubelet inside a container, or do you use the image to package it easily?

**Comment by LorbusChris** [comment]
@vrutkovs @haircommander it looks to me as though we are running the kubelet through the hyperkube binary (see https://github.com/openshift/installer/blob/master/data/data/bootstrap/systemd/units/kubelet.service.template#L15-L16),
which is extracted from the payload and written to disk during install (lives in https://github.com/openshift/kubernetes/blob/master/openshift-hack/images/hyperkube/Dockerfile.rhel).
I suppose crio could be distributed the same way, that is not as an RPM, but as a container.

**Comment by vrutkovs** [comment]
>that is not as an RPM, but as a container.

Yes, and we had this option in 3.x days. AFAIK some distros (Rancher) do that too - but node SIG doesn't officially support running kubelet (and container engine) in container, as it complicates volumes.

Perhaps we should be extracting binaries to `/usr/local/bin`?

**Comment by dghubble** [comment]
If we're just sideloading `cri-o` on our own with a "works on OKD with our RPMs" promise, as helpful as that is, it reduces the value prop for FCOS a bit. Could we at least have some "official" flow that guarantees `cri-o` can be installed successfully on FCOS? Ideally in FCOS that'd be an Ignition mechanism, but Dusty's script still seems to be the best approach that installs it currently. And some comittment to `cri-o` being a supported case on FCOS, beyond a question in the forums?

@haircommander Kubelet in a container [with podman](https://github.com/poseidon/typhoon/blob/master/aws/fedora-coreos/kubernetes/workers/fcc/worker.yaml#L36), like other on-host services (e.g. etcd).
@LorbusChris hyperkube was deprecated upstream back in k8s v1.18, so I'm guessing that's something custom openshift is doing

With v1.22, container runtime experience might become a differentiator for users picking from base OSes. I want both to be strong choices and don't weigh in on my users choice, but (aluding to a previous thread) carrot and stick can't work if the competition is passing out carrots.

**Comment by LorbusChris** [comment]
@dghubble https://github.com/coreos/rpm-ostree/issues/1435 would probably be the cleanest solution to this then

**Comment by haircommander** [comment]
I agree @LorbusChris , I think that's the best not-hacky way to guarantee cri-o can be installed (and kubelet can be installed with a corresponding version). we could even couple a kubernetes module with kubelet, cri-o and crictl

**Comment by dghubble** [comment]
> This is the main reason why cri-o is not shipped as part of the base FCOS. The daemon isn't a generic container runtime that can be freely upgraded by the OS, but it is instead interlocked with the higher level application/service (k8s control plane).

How about Fedora CoreOS shipping `containerd` then, as the general container runtime? Which would give a window of compatibility @lucab

**Comment by haircommander** [comment]
> > This is the main reason why cri-o is not shipped as part of the base FCOS. The daemon isn't a generic container runtime that can be freely upgraded by the OS, but it is instead interlocked with the higher level application/service (k8s control plane).
> 
> How about Fedora CoreOS shipping `containerd` then, as the general container runtime? Which would give a window of compatibility @lucab

What reasons are there for shipping a general container runtime? If the modules thing is figured out, and cri-o can be installed with kubernetes seamlessly, I don't see a need for containerd

**Comment by dghubble** [comment]
Some of the original value behind Container Linux being packageless were shipping a minimal OS suitable for cluster uses cases (i.e. container runtime is "new enough"). Adding RPMs is more a step toward traditional approaches and slower cadence. What happens when Kubernetes has a release, but cri-o doesn't have an RPM yet if they're in lock-step? For example, how would we test Kubernetes v1.21-beta.1 right *now*? We'd need to release without Fedora CoreOS, and I can see users gravitating toward Flatcar/containerd if this just isn't even a factor there.

To be clear, I don't care which container runtime is chosen. I have no horse in this race (thank you CRI). Just that it work well in these cases going forward.

**Comment by haircommander** [comment]
> Some of the original value behind Container Linux being packageless were shipping a minimal OS suitable for cluster uses cases (i.e. container runtime is "new enough"). Adding RPMs is more a step toward traditional approaches and slower cadence. What happens when Kubernetes has a release, but cri-o doesn't have an RPM yet if they're in lock-step? For example, how would we test Kubernetes v1.21-beta.1 right _now_? We'd need to release without Fedora CoreOS, and I can see users gravitating toward Flatcar/containerd if this just isn't even a factor there.

I see this as mostly a syncing of requirements. Up until now, we haven't had a request to test a yet-to-be-release cri-o in FCOS. I see no reason we couldn't ship the module early--assuming we can set expectation about stability before the .0 release. I'm happy to work together to setup cri-o packaging better on FCOS

**Comment by vrutkovs** [comment]
CRI-O is also being built upstream in non-module RPMs, rpm-ostree can install those - see [crio 1.20](https://build.opensuse.org/project/show/devel:kubic:libcontainers:stable:cri-o:1.20)

**Comment by jlebon** [comment]
> If we're just sideloading `cri-o` on our own with a "works on OKD with our RPMs" promise, as helpful as that is, it reduces the value prop for FCOS a bit. Could we at least have some "official" flow that guarantees `cri-o` can be installed successfully on FCOS? Ideally in FCOS that'd be an Ignition mechanism, but Dusty's script still seems to be the best approach that installs it currently.

Yes, this is #681. We definitely need to improve the UX on package layering. (Modularity is something that this sugar will need to consider as well.)

> And some comittment to `cri-o` being a supported case on FCOS, beyond a question in the forums?

Agreed we need to discuss this (IMO, yes this should be supported). Essentially, I think we need to:
- decide on a version (or range of versions) of k8s we want to support
- work with the cri-o team to have the matching RPMs for those versions available (either as proper modules in Fedora, or e.g. using [the new OS extensions work](https://github.com/coreos/rpm-ostree/pull/2439) to ship them in a side yum repo)
- add CI tests that sanity-checks the supported cri-o versions (at least with Prow, we can do full e2e testing for the version currently targeted by OKD)
- make sure that the install layer (OKD/Typhoon) knows how to drive rpm-ostree to install the right cri-o version for the target k8s version being installed (e.g. enable a yum repo, then call `rpm-ostree install cri-o-1.20`)

I don't think rpm-ostree natively supporting modules is a blocker for all this, but it definitely would make it easier.

**Comment by haircommander** [comment]
> > If we're just sideloading `cri-o` on our own with a "works on OKD with our RPMs" promise, as helpful as that is, it reduces the value prop for FCOS a bit. Could we at least have some "official" flow that guarantees `cri-o` can be installed successfully on FCOS? Ideally in FCOS that'd be an Ignition mechanism, but Dusty's script still seems to be the best approach that installs it currently.
> 
> Yes, this is #681. We definitely need to improve the UX on package layering. (Modularity is something that this sugar will need to consider as well.)
> 
> > And some comittment to `cri-o` being a supported case on FCOS, beyond a question in the forums?
> 
> Agreed we need to discuss this (IMO, yes this should be supported). Essentially, I think we need to:
> 
>     * decide on a version (or range of versions) of k8s we want to support
> 

I think supporting the three releases k8s upstream supports should work

>     * work with the cri-o team to have the matching RPMs for those versions available (either as proper modules in Fedora, or e.g. using [the new OS extensions work](https://github.com/coreos/rpm-ostree/pull/2439) to ship them in a side yum repo)

I am happy to go forward with either, though my preference is proper module support (it feels more idiomatic for fedora)

>     * add CI tests that sanity-checks the supported cri-o versions (at least with Prow, we can do full e2e testing for the version currently targeted by OKD)

big +1

> make sure that the install layer (OKD/Typhoon) knows how to drive rpm-ostree to install the right cri-o version for the target k8s version being installed (e.g. enable a yum repo, then call rpm-ostree install cri-o-1.20)

I suppose if we go the extensions route, we could use the existing OBS infrastructure for this (enable the OBS repo instead of some special cri-o one)

**Comment by jlebon** [comment]
Related: https://github.com/openshift/os/issues/498

**Comment by bgilbert** [comment]
> What reasons are there for shipping a general container runtime? If the modules thing is figured out, and cri-o can be installed with kubernetes seamlessly, I don't see a need for containerd

Fedora CoreOS is meant to be a general platform for running containers.  If there's a container runtime that people want to use, and there are no technical blockers to shipping it, we should probably ship it.  We already ship Moby and podman, and would already be shipping cri-o if not for the version-skew issue.

Is there some reason we shouldn't ship containerd?

**Comment by lucab** [comment]
By the way there is already a `containerd` [package](https://github.com/coreos/fedora-coreos-config/blob/257c8970b238d63113a2feb5eb9043e248e27e29/manifest-lock.x86_64.json#L132-L134) shipped in FCOS, which is getting in as a dependency. We don't cover that explicitly in our CI, so I don't know whether it actually works.

**Comment by rhatdan** [comment]
We need to look to shrink the size of FCOS not increase its size.

**Comment by bgilbert** [comment]
@rhatdan Sure; see also #186.  In general, we try to be [careful](https://github.com/coreos/fedora-coreos-tracker/issues/new?assignees=&labels=kind%2Fenhancement&template=new-package.md&title=New+Package+Request%3A+%3Cpackage+name%3E) about adding packages to the distro.  Historically, though, when a package is important to the core functionality of FCOS (supporting hardware and running containers) and can't be run in a container itself, we've tended to add it to the distro.  I suspect the biggest reductions in distro size would come from trimming dependencies from packages in our core set.

@dghubble From your perspective, what is FCOS missing in order to support containerd as a viable option?  Does the package mentioned in https://github.com/coreos/fedora-coreos-tracker/issues/767#issuecomment-802674793 meet your needs?

**Comment by rhatdan** [comment]
I want to replace all moby-engine/containerd with podman-docker, and then treat moby-engine/containerd the same way we treat kublet/cri-o as layers on top of FCOS.

**Comment by rhatdan** [comment]
du -sm /usr/bin/docker /usr/bin/dockerd /usr/bin/runc /usr/bin/containerd
69	/usr/bin/docker
112	/usr/bin/dockerd
19	/usr/bin/runc
55	/usr/bin/containerd

**Comment by rhatdan** [comment]
du -s /usr/bin/docker /usr/bin/dockerd /usr/bin/runc /usr/bin/containerd
69920	/usr/bin/docker
113696	/usr/bin/dockerd
19276	/usr/bin/runc
55772	/usr/bin/containerd

versus

du -s /usr/bin/podman /usr/bin/crun /usr/bin/conmon 
44580	/usr/bin/podman
356	/usr/bin/crun
132	/usr/bin/conmon

**Comment by dghubble** [comment]
I appreciate the efforts to make the container tools you work on smaller. Though the size of the binary has not been a decider for me in choosing a container runtime.

Providing container runtime(s) is a core part of choosing Fedora CoreOS and we need ready options to replace dockershim. If either containerd or cri-o or both here shipped that'd be nice. Both would keep the projects honest (why not compete head to head) and give options that are definitely in FCOS's wheelhouse. I could see maintenance pushback being legitimate.

@bgilbert I noticed containerd shipped with docker in December and tried it then. There were some difficulties getting the right `/etc/containerd/config.toml` (the default is junk values) for CNI to be happy and I stopped looking at it since it didn't seem officially supported anyway. I could revisit. https://github.com/poseidon/typhoon/pull/959

**Comment by rhatdan** [comment]
The problem with supporting all of the container work flows is size. Eliminating moby-engine and friends eliminates almost 20% of the size of FCOS. Podman can provide everthing you need to do for Docker.  

If you want to run Kubernetes on FCOS you are going to need more then just containerd.  You will need crictl, kublet, plus a few other tools, these are not shipped by default in FCOS and the Kublet tends to be tied to specific versions of the container engine.

**Comment by dghubble** [comment]
@rhatdan In this issue, I'm interested in what Kubernetes compatible container runtimes will be supported on FCOS. I'm already using and happy with podman as the container runner on-host for systemd units, etc. I don't think we have to rehash that topic (which I view as separate) or make it about podman vs docker. I'm already sold on podman.

You can run conformant Kubernetes on FCOS today, without layerying in any RPM packages. Kubelet was covered [here](https://github.com/coreos/fedora-coreos-tracker/issues/767#issuecomment-797636475). Its an Openshift specific design choice to use RPMs (which is fine, I don't want to make this about how y'all build your product). Just to emphasize that this pattern is a choice, not a neccessity. Today the base OS provides us the container runtime for Kubernetes, currently docker(shim), and in future some suitable replacement.

**Comment by cgwalters** [comment]
Soon one will be able to do (via e.g. Ignition):
```
$ sed -i s,enabled=,enabled=1, /etc/yum.repos.d/fedora-modular.repo
$ rpm-ostree install --apply-live crio
```

That will pull the latest crio; indeed rpm-ostree is unaware of modules though and will just pick the latest.

**Comment by jlebon** [comment]
We could ask rpm-ostree to pull a specific version but that gets messed up by release field (e.g. `rpm-ostree install cri-o-1.19` won't work). One pretty easy hack around (and I swear I'm not trying to get out of rpm-ostree supporting modularity properly, but it could be a viable short-term solution) is to have the cri-o modules have a symbolic `Provides` that's just `Provides: cri-o-%{version} = %{release}` and then we can just do `rpm-ostree install cri-o-1.19` and make sure the host stays on the 1.19 stream.

**Comment by wernerb** [comment]
> I appreciate the efforts to make the container tools you work on smaller. Though the size of the binary has not been a decider for me in choosing a container runtime.
> 
> Providing container runtime(s) is a core part of choosing Fedora CoreOS and we need ready options to replace dockershim. If either containerd or cri-o or both were shipped that'd be nice. Both would keep the projects honest (why not compete head to head) and give options that are definitely in FCOS's wheelhouse. I could see maintenance pushback being legitimate though.
> 
> @bgilbert I noticed containerd shipped with docker in December and tried it then. There were some difficulties getting the right `/etc/containerd/config.toml` (the default is junk values) for CNI to be happy and I stopped looking at it since it didn't seem officially supported anyway. I could revisit. [poseidon/typhoon#959](https://github.com/poseidon/typhoon/pull/959)
> 
> EDIT: It kinda works minimally. But it needs additional CNI plugins it didn't before (firewall, tuning, etc). Sideloading those by hand (which could later be done by the flannel-cni daemonset) and pods can start at least.

The CNI plugins are shipped in fcos but most CNI implementations require writing new binaries and reading plugins from `/opt/cni/bin`. They are located in `/usr/libexec/cni` which are readonly. 
I solved this by symlinking each cni plugin binary with:
```
ExecStartPre=-/bin/sh -c "for f in /usr/libexec/cni/*; do ln -s \"$f\" /opt/cni/bin/$(basename $f); done"
```

> It would be helpful to have `crictl`.

`crictl` is crucial for debugging workloads previously able to use `docker`.

**Comment by dghubble** [comment]
@wernerb I don't see CNI plugins as being on Fedora CoreOS's plate. CNI plugins are often placed on hosts by DaemonSets. Not a worry.
@bgilbert I'll post back with more details, probably not during a work week
@cgwalters thanks, will see if we can get this into a systemd unit and try it out

**Comment by Conan-Kudo** [comment]
> We could ask rpm-ostree to pull a specific version but that gets messed up by release field (e.g. `rpm-ostree install cri-o-1.19` won't work)

To note, DNF does support `dnf install 'foo = version'` as a way to request package installation. Does rpm-ostree not have that capability?

**Comment by jlebon** [comment]
> > We could ask rpm-ostree to pull a specific version but that gets messed up by release field (e.g. `rpm-ostree install cri-o-1.19` won't work)
> 
> To note, DNF does support `dnf install 'foo = version'` as a way to request package installation. Does rpm-ostree not have that capability?

Sorry, I mixed up terminology in that comment. The issue isn't the release field, it's the patch component of the version string. E.g. you can't ask for `cri-o-1.20`. You have to pin to a specific version but here we'd want to pin at the minor level.

Anyway, I've been working on teaching modules to rpm-ostree so hopefully soon we can do this properly without any hacks.

**Comment by buckaroogeek** [comment]
In my home lab I use fedora 33 to host a small kubernetes cluster with cri-o. I use dnf modularity for cri-o version management and the dnf versionlock plugin to provide version management of packages from the upstream kubernetes repository. `dnf versionlock add kube???-1.20.?` will allow dnf to install the current version kubeadm, kubectl, and kubelet. It also will allow updates to any patch version of the 1.20 release. Would it be feasible to add a similar capability to rpm-ostree?

**Comment by dghubble** [comment]
I've been running clusters using `containerd` (available by default) to replace the `docker-shim` lately. With the broad use, forward compatability, and already being installed, this is the likely direction for the underlying container runtime in Typhoon.

```
OS-IMAGE                          KERNEL-VERSION             CONTAINER-RUNTIME
Fedora CoreOS 34.20210529.2.0     5.12.7-300.fc34.x86_64     containerd://1.5.0
Fedora CoreOS 34.20210529.2.0     5.12.7-300.fc34.x86_64     containerd://1.5.0
Fedora CoreOS 33.20210413.dev.0   5.10.19-200.fc33.aarch64   containerd://1.4.4
```

cc @bgilbert

**Comment by anthr76** [comment]
For users wishing to use cri-o this leaves us in a sad place :( 

Basically one of the most appealing options is Kubic though using it in a automated fashion with tools like mathcbox and typoon is incredibly hard with AutoYasT. It's a great opportunity for FCOS to pickup cri-o and have it as a default though I understand the technical constraints..

**Comment by jlebon** [comment]
Update on this: proper support for modularity has now merged in rpm-ostree (https://github.com/coreos/rpm-ostree/pull/2760). So in the next release, one should be able to do e.g.

```
$ rpm-ostree module install cri-o:1.20
```

For now, it'd work to do this in a systemd unit and reboot like in https://docs.fedoraproject.org/en-US/fedora-coreos/os-extensions/. But eventually, we still want to polish the UX for extensions as in https://github.com/coreos/fedora-coreos-tracker/issues/681.

Related patch to stop disabling modular repos in FCOS at: https://github.com/coreos/fedora-coreos-config/pull/1149.

**Comment by jlebon** [comment]
So circling back to earlier discussions here, leveraging modularity I think at this point we should be able to form a stance on which supported cri-o runtime versions are supported in collaboration with the containers team.

@haircommander, you mentioned in https://github.com/coreos/fedora-coreos-tracker/issues/767#issuecomment-799700202 we should just support all the versions supported upstream, which makes sense to me. Should we also have a stream for the *next* development version? (E.g. right now `cri-o:1.22`.)

Then the next step would be adding CI for testing the supported cri-o versions.

**Comment by haircommander** [comment]
In my ideal universe, we would package all the k8s binaries, cri-o and crictl as part of a kubernetes package, all of which would be modular per-release

(another step we need is reintroducing cri-o package to fedora--there was a mistake and it was orphaned:  https://bugzilla.redhat.com/show_bug.cgi?id=1970050

**Comment by jlebon** [comment]
@haircommander, I noticed that f35 only has 1.19, while f34 has 1.20. And f36 only has 1.21. Is this intended or does the module just need some loving?

**Comment by Conan-Kudo** [comment]
Some effort needs to be put in to bring everything back in sync.

**Comment by Conan-Kudo** [comment]
> In my ideal universe, we would package all the k8s binaries, cri-o and crictl as part of a kubernetes package, all of which would be modular per-release
> 

The kubernetes module should probably have both kubernetes and cri-o. No more separate cri-o module.

**Comment by dghubble** [comment]
@Conan-Kudo some of us would only want cri-o. Not the rest.

**Comment by navaati** [comment]
@dghubble As I understand things it's just a module so you can enable it, only install the cri-o package and don't install the kube package and you won't get anything you don't want, I think it's ok ?

**Comment by Conan-Kudo** [comment]
Yes, indeed. You can effectively treat a module as a subrepo.

**Comment by haircommander** [comment]
> @haircommander, I noticed that f35 only has 1.19, while f34 has 1.20. And f36 only has 1.21. Is this intended or does the module just need some loving?

I don't think I understand how to use modules well haha. It needs some love and some research on my part :) 

> The kubernetes module should probably have both kubernetes and cri-o. No more separate cri-o module.

@Conan-Kudo I like the sound of that! Do you think it is still useful to have the `cri-o` module? I can see it making sense to mirror the two if that wasn't too difficult

**Comment by Conan-Kudo** [comment]
The two ways it could be done:

* Kubernetes module includes kubernetes and cri-o, and you can have a separate cri-o module if you'd like for just cri-o
* Kubernetes module includes only kubernetes, but depends on separate cri-o module.

The second option makes sense to do, actually, if we expect kubernetes and cri-o updates to happen independently. Having them in one module basically forces that both are built together every time, even if you're only updating one of them.

**Comment by anthr76** [comment]
I would be a fan of the second approach personally. Can the cri-o dependency be automatically resolved or will the module need to enabled first?

**Comment by buckaroogeek** [comment]
Speaking from a cri-o and kubernetes consumer perspective (of the fedora-based home lab variety), I recommend separate modules at this point. CRI-O is still in the incubating stage of adoption (according to cri-o web site) so combining into a unified module might be premature - even though a module can essentially function as a sub repo. Also, cri-o and kubernetes are intended to synchronize on the minor version but patch levels can be different due to the differing release policies of both products.  

@haircommander - I would appreciate getting the cri-o module for F34 and F35 caught up to the current release.  I know everyone is over committed with tasks however.

thanks

**Comment by haircommander** [comment]
> Kubernetes module includes kubernetes and cri-o, and you can have a separate cri-o module if you'd like for just cri-o

the advantage of this approach is kubernetes package could be modular based on the minor version of kubernetes, and would come with the corresponding cri-o/crictl version. Having them be separate would involve the user correctly installing the correct module.

Alternatively, CRI-O could specify:
`Provides: kubernetes-cri, kubernetes-cri-1.22`
and the kubernetes 1.22 release could rely on `kubernetes-cri-1.22` or something

**Comment by Conan-Kudo** [comment]
The problem is that kubernetes can be used with other CRIs, and we don't currently have our kubernetes package configured to default to CRI-O yet. If we did, then we can do something like `Provides: kubernetes(cri) = 1.22` and have the Kubernetes package `Requires: (kubernetes(cri) = 1.22 or containerd or moby-engine or docker or docker-ce)`

**Comment by buckaroogeek** [comment]
DNF has a version lock plug-in which can be useful in this type of scenario. The plug-in can be used to block minor level changes but allow patch level changes to a target rpm (or set of rpms).

**Comment by Conan-Kudo** [comment]
> DNF has a version lock plug-in which can be useful in this type of scenario. The plug-in can be used to block minor level changes but allow patch level changes to a target rpm (or set of rpms).

We should not rely on `dnf versionlock`, especially since rpm-ostree doesn't have that capability.

**Comment by jlebon** [comment]
Was chatting with @LorbusChris and @haircommander about this. We all think it would be best to keep the kubernetes and cri-o modules separate. @haircommander will keep the cri-o module more up-to-date and match the N-2 support cadence of upstream.

For CI, we're thinking both:
- Hook up OKD to f-c-c using Prow. OKD will move to using the cri-o module.
- Hook up cri-o integration suite to f-c-c using CoreOS CI or Prow. These are single host tests, so we'd just need to bring up a VM per stream version.

**Comment by dghubble** [comment]
Checking in about `cri-o` and its status.

* rpm-ostree *should* support modules (https://github.com/coreos/rpm-ostree/pull/2760)
* `rpm-ostree module install cri-o:1.20` is an outdated command today (Unknown command 'module' @jlebon)
* Upstream tests [curl down scripts](https://github.com/cri-o/cri-o/blob/main/scripts/node_e2e_installer#L18) (non-starter)
* Use of modules still seems in-progress (https://github.com/coreos/fedora-coreos-tracker/issues/88, https://github.com/coreos/fedora-coreos-tracker/issues/681)
* Test coverage was dropped (https://github.com/coreos/fedora-coreos-config/pull/1211)
* Some imperative commands were found:
  * sudo rpm-ostree ex module install cri-o:1.20/default (works, but pretty old)
  * sudo rpm-ostree ex module install cri-o:1.21/default (fails)
  * sudo rpm-ostree ex module install cri-o:1.22/default (fails)
* Open [bug reports](https://bugzilla.redhat.com/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&classification=Fedora&component=cri-o&product=Fedora%20Modules) about `modules/cri-o` not able to be installed

Docs:

* FCOS docs don't mention modules (https://github.com/coreos/fedora-coreos-docs)
* Modularity docs don't mention FCOS (https://docs.fedoraproject.org/en-US/modularity/)
* cri-o docs don't mention FCOS (https://cri-o.io/)
* container/cri-o still shows using `atomic` (https://src.fedoraproject.org/container/cri-o)
* modules/cri-o shows no directions (https://src.fedoraproject.org/modules/cri-o/tree/1.22)

The desired end state here is an official Butane config to install a specific version of `cri-o` reliably.

**Comment by jlebon** [comment]
> Checking in about `cri-o` and its status.
> 
> * rpm-ostree _should_ support modules ([Add support for modules rpm-ostree#2760](https://github.com/coreos/rpm-ostree/pull/2760))
> 
> * `rpm-ostree module install cri-o:1.20` is an outdated command today (Unknown command 'module' @jlebon)

Sorry about that, my example command in https://github.com/coreos/fedora-coreos-tracker/issues/767#issuecomment-892119669 was missing the `ex`. I've added that now.

> * Use of modules still seems in-progress ([Modularity for extending the host #88](https://github.com/coreos/fedora-coreos-tracker/issues/88)

Thanks, that issue is obsolete IMO. I closed it.

 > * Test coverage was dropped ([tests/misc-ro: drop `rpm-ostree ex module install` test fedora-coreos-config#1211](https://github.com/coreos/fedora-coreos-config/pull/1211))

It was moved to the rpm-ostree CI: https://github.com/coreos/rpm-ostree/blob/f3a977a32bc76c0776f7d12e5c869d46fda86e9f/tests/kolainst/destructive/layering-modules#L13-L21

Related to this, we're also working on adding k8s e2e tests (which uses cri-o) in https://github.com/coreos/fedora-coreos-config/pull/1297. It doesn't use the Fedora module, but contributes to providing confidence that newer versions won't break.

> * Some imperative commands were found:      
>     * sudo rpm-ostree ex module install cri-o:1.20/default (works, but pretty old)
>     * sudo rpm-ostree ex module install cri-o:1.21/default (fails)
>     * sudo rpm-ostree ex module install cri-o:1.22/default (fails)

Yes, this is an issue (mentioned it [higher up too](https://github.com/coreos/fedora-coreos-tracker/issues/767#issuecomment-917191270)). We chatted recently with @haircommander who's working on updating the module.

> * FCOS docs don't mention modules ([coreos/fedora-coreos-docs](https://github.com/coreos/fedora-coreos-docs))

Yeah, agreed we should add docs there, but sadly I don't think we're quite ready for it yet. I'd like to have updated modules and more tests in CI first.

> * Modularity docs don't mention FCOS ([docs.fedoraproject.org/en-US/modularity](https://docs.fedoraproject.org/en-US/modularity/))

Modularity is a Fedora effort and not variant-specific, so it wouldn't mention FCOS there.

> The desired end state here is an official Butane config to install a specific version of `cri-o` reliably.

Yes, agreed. :+1: 

As you've noticed, there's still a lot of work left here. Thank you for highlighting this!

**Comment by buckaroogeek** [comment]
I note that `sudo rpm-ostree ex module install cri-o:1.21` now works but 1.22 still fails. Fedora 34 only has cri-o 1.21 and older available as modules, whereas Fedora 35 also includes version 1.22.

**Comment by Conan-Kudo** [comment]
What? Of course Fedora Linux 34 has Go 1.16. We updated to Go 1.16 in F34: https://fedoraproject.org/wiki/Changes/golang1.16

**Comment by dghubble** [comment]
On F35 too, there are still problems with a manual install of `crio` `1.22`.

```
NAME="Fedora Linux"
VERSION="35.20211029.3.0 (CoreOS)"
```

```
sudo rpm-ostree ex module install cri-o:1.22
NOTICE: Experimental commands are subject to change.
Checking out tree ccdac24... done
Enabled rpm-md repositories: fedora-cisco-openh264 fedora-modular updates-modular updates fedora updates-archive
Importing rpm-md... done
rpm-md repo 'fedora-cisco-openh264' (cached); generated: 2021-09-21T18:07:30Z solvables: 4
rpm-md repo 'fedora-modular' (cached); generated: 2021-10-29T10:04:16Z solvables: 1283
rpm-md repo 'updates-modular' (cached); generated: 2021-12-01T01:40:40Z solvables: 1328
rpm-md repo 'updates' (cached); generated: 2021-12-06T01:00:02Z solvables: 12664
rpm-md repo 'fedora' (cached); generated: 2021-10-29T10:17:40Z solvables: 65732
rpm-md repo 'updates-archive' (cached); generated: 2021-12-06T01:15:12Z solvables: 13006
error: Installing modules: Problems appeared for module install request:
  - Modular dependency problem: Problem during install for module 'cri-o' stream '1.22': No default profile found for cri-o:1.22:3520211111201912:f27b74a8:x86_64
```

**Comment by buckaroogeek** [comment]
The steps I took on FCOS f35 for cri-o was:

sudo rpm-ostree ex module enable cri-o:1.22
sudo rpm-ostree install cri-o

**Comment by dustymabe** [comment]
Hey @dghubble - we identified this problem last week: in the cri-o module there is a profile called `default` but it's not explicitly marked as the default so you end up with the problem you observed. We're working with @haircommander to get the `default` profile marked as default.

For now you can do something like what @buckaroogeek mentioned above or in one command:

```
[core@cosa-devsh ~]$ sudo rpm-ostree ex module install cri-o:1.22/default
NOTICE: Experimental commands are subject to change.
Checking out tree 13af377... done
Enabled rpm-md repositories: fedora-cisco-openh264 fedora-modular updates-modular updates fedora updates-archive
Importing rpm-md... done
rpm-md repo 'fedora-cisco-openh264' (cached); generated: 2021-09-21T18:07:30Z solvables: 4
rpm-md repo 'fedora-modular' (cached); generated: 2021-10-29T10:04:16Z solvables: 1283
rpm-md repo 'updates-modular' (cached); generated: 2021-12-01T01:40:40Z solvables: 1328
rpm-md repo 'updates' (cached); generated: 2021-12-07T01:51:30Z solvables: 12820
rpm-md repo 'fedora' (cached); generated: 2021-10-29T10:17:40Z solvables: 65732
rpm-md repo 'updates-archive' (cached); generated: 2021-12-07T02:18:41Z solvables: 13149
Resolving dependencies... done
Will download: 2 packages (34.6?MB)
Downloading from 'updates-modular'... done
Importing packages... done
Checking out packages... done
Running pre scripts... done
Running post scripts... done
Running posttrans scripts... done
Writing rpmdb... done
Writing OSTree commit... done
Staging deployment... done
Added:
  cri-o-1.22.1-1.module_f35+13370+fa0e2e9f.x86_64
  cri-tools-1.22.0-3.module_f35+13370+fa0e2e9f.x86_64

```

**Comment by jlebon** [comment]
> For now you can do something like what @buckaroogeek mentioned above or in one command:

Note that the semantics are not the same between these two approaches. `rpm-ostree ex module install` installs profiles, which is a set of packages not necessarily dependent on each other. On the other hand, `rpm-ostree ex module enable` only enables a stream and `rpm-ostree install` may install any packages from that stream; only its dependencies will be pulled in.

As a concrete example, the steps from @buckaroogeek will not pull in `cri-tools` since it's not a strict dependency of `cri-o`, while those from @dustymabe will since it's part of the `default` profile.

**Comment by Richterrettich** [comment]
Does this also imply that in the future, the preferred method of installation for Kubernetes is module based too? Otherwise, it would not easily be possible to keep crio/kubernetes versions in sync.

**Comment by LorbusChris** [comment]
I agree that a Kubernetes module would be the sanest approach to deal with multiple versions here. I think @janch32 is the maintainer of the RPM in Fedora today. It's probably best to work with him to get that RPM modularized.

**Comment by LorbusChris** [comment]
Sorry, I tagged the Jan (my apologies!). The one I meant should be @ingvagabund

**Comment by buckaroogeek** [comment]
There is an existing, but orphaned, module package for Kubernetes in the fedora package management system. I ran across it while exploring k8 installation on CoreOS using fedora rpms vs upstream rpms. Given the current support policy from kubernetes (the last 3 minor releases - so 1.21, 1.22, 1.23 today for example) and the number of patch releases in each minor release, a modular approach would be very useful.

**Comment by dghubble** [comment]
Followup on https://github.com/coreos/fedora-coreos-tracker/issues/767#issuecomment-967793379
> The desired end state here is an official Butane config to install a specific version of cri-o reliably

We still lack a declarative butane config for getting `cri-o`. Kubernetes v1.24 will remove `dockershim`.

Manual commands (which we don't ultimately want to resort to) don't seem ready (Kubernetes v1.23 came out 31 days ago).

```
sudo rpm-ostree ex module enable cri-o:1.23                               
NOTICE: Experimental commands are subject to change.
Checking out tree 30c82ee... done
Enabled rpm-md repositories: fedora-cisco-openh264 fedora-modular updates-modular updates fedora updates-archive
Importing rpm-md... done
rpm-md repo 'fedora-cisco-openh264' (cached); generated: 2021-09-21T18:07:30Z solvables: 4
rpm-md repo 'fedora-modular' (cached); generated: 2021-10-26T05:08:36Z solvables: 1283
rpm-md repo 'updates-modular' (cached); generated: 2022-01-06T01:57:22Z solvables: 1371
rpm-md repo 'updates' (cached); generated: 2022-01-08T01:04:36Z solvables: 16018
rpm-md repo 'fedora' (cached); generated: 2021-10-26T05:31:27Z solvables: 65732
rpm-md repo 'updates-archive' (cached); generated: 2022-01-08T01:31:50Z solvables: 18937
error: Problems appeared for module enable request:
  - Unable to resolve argument 'cri-o:1.23'

less /etc/os-release 
NAME="Fedora Linux"
VERSION="35.20211215.3.0 (CoreOS)"
```

**Comment by dustymabe** [comment]
hmm. There's activity on the [`1.23` branch](https://src.fedoraproject.org/modules/cri-o/commits/1.23), but I don't see any releases of the module. @haircommander - can you put out a release?

**Comment by LorbusChris** [comment]
Go 1.17 won't land in Fedora 35, so there's likely not going to be an official release of cri-o 1.13 until F36 :(

**Comment by Conan-Kudo** [comment]
Well, modular Go can be used to build `cri-o:1.23` module on F35, couldn't it?

**Comment by LorbusChris** [comment]
It could, yes, but right now there's not Golang module being maintained for Fedora

**Comment by ingvagabund** [comment]
Hello everyone

> I agree that a Kubernetes module would be the sanest approach to deal with multiple versions here. I think @janch32 is the maintainer of the RPM in Fedora today. It's probably best to work with him to get that RPM modularized.

Unfortunately, I am no longer active in the RPM Fedora area. The [Kubernetes](https://src.fedoraproject.org/rpms/kubernetes) rpms in Fedora has been lacking maintenance for some time. I recall there was an effort to make Kubernetes modular. However, I was not part of the effort. If anyone is interested in proceeding here and keeping the Kubernetes rpms up2date, I can share the access rights to make so.

**Comment by travier** [comment]
Commands for `1.22`: https://github.com/coreos/fedora-coreos-config/pull/1406

**Comment by anthr76** [comment]
> Hello everyone
> 
> > I agree that a Kubernetes module would be the sanest approach to deal with multiple versions here. I think @janch32 is the maintainer of the RPM in Fedora today. It's probably best to work with him to get that RPM modularized.
> 
> Unfortunately, I am no longer active in the RPM Fedora area. The [Kubernetes](https://src.fedoraproject.org/rpms/kubernetes) rpms in Fedora has been lacking maintenance for some time. I recall there was an effort to make Kubernetes modular. However, I was not part of the effort. If anyone is interested in proceeding here and keeping the Kubernetes rpms up2date, I can share the access rights.

I have not worked directley with Fedora RPMs but familiar of the process this is something I would be interested in. I'm on matrix if you wanted to speak directley @anthr76:mozilla.org

**Comment by buckaroogeek** [comment]
Sure I, for one, can give it a go.

:)

Actually, yes I am very interested in helping. Reviewing the package maintainer web pages now. And will follow through on the needed steps.  I might also suggest helm as that seems to also be lagging in packaging support.  It would seem that getting some solid traction around go and a go module would be a first step? After getting the package maintainers introductions done.

**Comment by dustymabe** [comment]
> @haircommander
> So it sounds like our eventual ideal situation is having modular go, cri-o and kubernetes.

It looks like the golang maintainers started a discussion about golang and different versions [recently](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/LUMQLWNXRWY2JYRJQO6VI3JIDN5QUR5H/) on Fedora Devel. It looks like the consensus there was to use golang compat packages (not a module) if they wanted to support different versions in Fedora. Which IIUC should be fine for what we need.

I think the status is:

- golang
    - work with maintainers to help them achieve compat packages
        - email asm@redhat.com to get involved
- kubernetes
    - create a kubernetes module.
        - work with @haircommander (pehunt@redhat.com) and @ingvagabund (jchaloup@redhat.com)
- crio
    - @haircommander should have this one covered.


@buckaroogeek @anthr76 - does that sound good?

**Comment by buckaroogeek** [comment]
@dustymabe - Yes, it does for me.

**Comment by Conan-Kudo** [comment]
I can help with the non-modular `kubernetes` and `cri-o` packages that are required to be present alongside modular ones.

**Comment by ingvagabund** [comment]
@buckaroogeek @anthr76 feel free to share you fedora usernames so I can add you as committers to the kubernetes project. Checking the previous commits might give you some hints on how to do update the kubernetes to the next version. I have not installed the kubernetes rpms for ages so it might be challenging at the beginning to make it work. Though learning from mistakes will eventually yield the right results.

**Comment by buckaroogeek** [comment]
@ingvagabund - buckaroogeek is the name I use in FAS. 

Thanks for the tips and assist. Will

**Comment by ingvagabund** [comment]
When adding you as a committer, I am getting `This user must be in one of the following groups to be allowed to be added to this project: packager`. I believe you need a sponsor. Not sure what's the process these days to obtain one.

**Comment by buckaroogeek** [comment]
@ingvagabund - I do not want to further side track this issue. May I contact you directly? A ticket needs to be filed with the package sponsor pagure instance by a current sponsor.

**Comment by ingvagabund** [comment]
> I do not want to further side track this issue

**Comment by ingvagabund** [comment]
> I do not want to further side track this issue. May I contact you directly?

@buckaroogeek @anthr76 what channel do you prefer? IRC, email, ...?

**Comment by anthr76** [comment]
> > I do not want to further side track this issue. May I contact you directly?
> 
> @buckaroogeek @anthr76 what channel do you prefer? IRC, email, ...?

I'm on matrix as @anthr76:mozilla.org I'm in #fedora-coreos and #fedora-golang which is also bridged with IRC

**Comment by buckaroogeek** [comment]
Also on matrix as buckaroogeek. Fedora-coresos and fedora-arm mostly

**Comment by Conan-Kudo** [comment]
@ingvagabund @anthr76 Please file a request for sponsoring @anthr76 through to co-maintain kubernetes here: https://pagure.io/packager-sponsors

**Comment by cgwalters** [comment]
I think this one is going to intersect with https://github.com/coreos/enhancements/blob/main/os/coreos-layering.md for many people.

**Comment by dustymabe** [comment]
From @jlebon's memory (which is much better than my memory) in the video community meeting today he think's the remaining steps here are:

- work with the current RPM maintainers (@anthr76 and friends) on how we should recommend our users consume the rpms.
- develop documentation around the recommendation
- add a test that exercises the recommendation

**Comment by buckaroogeek** [comment]
Please let me know how I can help.  At one point, Anthony (@anthr76) and I were considering modules for kubernetes, but the approach taken by the nodejs and haskell maintainers on multiple versions seems to be more approachable. I have been negligent in getting a change request submitted to outline the path forward.  

The version of go supported in a given Fedora release would shape the version(s) of kubernetes available from Fedora repositories.  For example, F37 has go 1.19.x. Kubernetes 1.25, 1.24, and 1.23 would be available allowing Fedora, including CoreOS, to function as a node for any supported  version of Kubernetes and as a workstation hosting kubectl to manage those clusters.

Regardless of the approach, I would like to see (and help produce) good documentation for end users around the installation of vanilla kubernetes on all suitable variants of Fedora.
